File: frontend/src/app/service/topic.service.ts
import {Injectable} from '@angular/core';
import {HttpClient, HttpParams} from '@angular/common/http';
import {Observable} from 'rxjs';

@Injectable({
  providedIn: 'root'
})
export class TopicService {

  constructor(private http: HttpClient) { }

  /**
   *
   * @param clusterIndex Cluster index
   */
  public getTopics(clusterIndex: number): Observable<{
    names: string[],
  }> {
    return this.http.get<{
      names: string[],
    }>(`/api/v1/cluster/${clusterIndex}/topics/info`);
  }

  /**
   *
   * @param clusterIndex Cluster index
   * @param topicName Topic name
   */
  getTopicInfo(clusterIndex: number, topicName: string): Observable<{
    partitions: Partition[]
  }> {
    return this.http.get<{
      partitions: Partition[]
    }>(`/api/v1/cluster/${clusterIndex}/topics/${topicName}/info`);
  }

  /**
   *
   * @param clusterIndex Cluster index
   * @param topicName Topic name
   * @param partition Partition index
   * @param start Start offset
   * @param count Messages count
   * @param direction Messages fetch direction
   */
  getMessages(clusterIndex: number, topicName: string, partition: PartitionIndex,
              start: number, count: number, direction: Direction = 'asc'): Observable<{
    recordList: KRecord[],
    partition: number,
    offset: number,
    count: number,
    lastTimestamp: number,
  }> {
    const params = new HttpParams()
      .set('direction', direction);

    return this.http.get<{
      recordList: KRecord[],
      partition: number,
      offset: number,
      count: number,
      lastTimestamp: number,
    }>(`/api/v1/cluster/${clusterIndex}/topics/${topicName}/messages/partition/${partition}/offset/${start}/count/${count}`, {
      params,
    });
  }
}

File: backend/src/main/java/com/ringcentral/kv/service/TopicService.java
package com.ringcentral.kv.service;

import com.ringcentral.kv.entity.Partition;
import com.ringcentral.kv.entity.TopicInfo;
import com.ringcentral.kv.entity.TopicsInfo;
import org.apache.avro.specific.SpecificRecordBase;
import org.apache.kafka.clients.consumer.Consumer;
import org.apache.kafka.common.PartitionInfo;
import org.apache.kafka.common.TopicPartition;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Map;
import java.util.Set;

@Service
public class TopicService {
    
    @Value("${poll.timeout}")
    private Long pollTimeout;

    @Autowired
    private KafkaClusterService kafkaClusterService;

    /**
     * Fetch all topics info
     * 
     * @param clusterIndex Cluster index
     * 
     * @return Topics info
     */
    public synchronized TopicsInfo getTopicsInfo(Integer clusterIndex) {
        String clusterHost = kafkaClusterService.getClusterHost(clusterIndex);

        Consumer<String, SpecificRecordBase> consumer = kafkaClusterService.createConsumer(clusterHost);

        Map<String, List<PartitionInfo>> topics = kafkaClusterService.getTopics(consumer, this);
        
        assert topics != null : "Topics doesn't fetched";

        Set<String> names = topics.keySet();

        TopicsInfo topicsInfo = new TopicsInfo();

        topicsInfo.setNames(names);

        return topicsInfo;
    }

    /**
     * Get topics info
     * 
     * @param clusterIndex Kafka cluster index
     * @param topicName Topic name
     *                  
     * @return Topic info
     */
    public TopicInfo getTopicInfo(Integer clusterIndex, String topicName) {
        String clusterHost = kafkaClusterService.getClusterHost(clusterIndex);
        
        TopicInfo topicInfo = new TopicInfo();

        Consumer<String, SpecificRecordBase> consumer = kafkaClusterService.createConsumer(clusterHost);

        List<TopicPartition> topicPartitions = kafkaClusterService.getTopicPartitions(topicName, consumer);
        
        assert topicPartitions != null : "Partitions doesn't fetched";

        consumer.assign(topicPartitions);

        List<Partition> partitionInfos = kafkaClusterService.getPartitionInfos(consumer, topicPartitions);

        topicInfo.setPartitions(partitionInfos);

        consumer.close();

        return topicInfo;
    }
}
